cmd/kubeadm/app/cmd/options/generic.go

```diff
@@ -54,6 +54,9 @@ func AddIgnorePreflightErrorsFlag(fs *pflag.FlagSet, ignorePreflightErrors *[]st
 
 // AddControlPlanExtraArgsFlags adds the ExtraArgs flags for control plane components
 func AddControlPlanExtraArgsFlags(fs *pflag.FlagSet, apiServerExtraArgs, controllerManagerExtraArgs, schedulerExtraArgs *map[string]string) {
+	// TODO: https://github.com/kubernetes/kubeadm/issues/1601
+	// Either deprecate these flags or handle duplicate keys.
+	// Currently the map[string]string returned by NewMapStringString() doesn't allow this.
```

Comment:
not sure what to do with this.
i think i'm +1 to deprecate these. they are a bit too complex for CLI flags - e.g. `--extra-args=foo1=bar1,foo2=bar2`

`--feature-gates` falls into the same bucket, but we actually lack config for it during upgrade, IIRC.
alternatively they need to be refactored to support the new []Arg type and not use map[string]string as the CLI flag backing type.

Reply:
i think i will just leave this TODO and keep the issue https://github.com/kubernetes/kubeadm/issues/1601 open for the time being.